### PR TITLE
nydus: refactor driver implementation using converter

### PR DESCRIPTION
**Pull request by imeoer**
Use converter package from nydus-snapshotter to refactor layer
conversion workflow.

The package converts OCI tar layer to nydus blob layer, and then
merges all blob layers into final bootstrap layer.

The benefit of this change is that there is no need for containerd
to unpack the OCI layer and mount it as a snapshot, which improves
the speed and stability of the image layer conversion.

### CODE REVIEW

**Comment by hsiangkao** [comment]
I think we'd better to have a preliminary analysis of https://github.com/dragonflyoss/image-service/issues/565 before merging this...

File: pkg/driver/nydus/convert.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+		tr, err := compression.DecompressStream(rdr)
+		if err != nil {
+			return nil, err
+		}
+
+		digester := digest.SHA256.Digester()
+		pr, pw := io.Pipe()
+		tw, err := nydusify.Pack(ctx, io.MultiWriter(pw, digester.Hash()), nydusify.PackOption{
+			RafsVersion: "6",
```
Comment (liubogithub) [inline pkg/driver/nydus/convert.go]: Is it supposed to be only version 6 is supported?

File: pkg/driver/nydus/utils/archive.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/archive/compression"
+	"github.com/opencontainers/go-digest"
+)
+
+// PackTargz makes .tar(.gz) stream of file named `name` and return reader
```
Comment (liubogithub) [inline pkg/driver/nydus/utils/archive.go:29]: returns

File: pkg/driver/nydus/utils/utils.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	return false
+}
+
+func IsNydusManifest(manifest *ocispec.Manifest) bool {
+	for _, layer := range manifest.Layers {
+		if layer.Annotations == nil {
+			continue
+		}
+		if layer.Annotations[LayerAnnotationNydusBootstrap] != "" {
```
Comment (liubogithub) [inline pkg/driver/nydus/utils/utils.go:86]: Is it the same with the following?
```
if _, ok :=layer.Annotations[LayerAnnotationNydusBootstrap]; ok {
}
```

File: misc/config/config.yaml.nydus.tmpl [hunk 4]
```diff
@@ -41,23 +41,24 @@ converter:
...
+
       # specify nydus format version, possible values: `5`, `6` (EROFS-compatible), default is `5`
-      rafs_version: 5
+      # rafs_version: 5
+
       # specify nydus blob compression algorithm, possible values: `none`, `lz4_block`, `gzip`, `zstd`, default is `lz4_block`
       # rafs_compressor: zstd
-      # ensure that both OCIv1 manifest and nydus manifest are present as manifest index in the target image.
-      # it's used for containerd to support running OCIv1 image or nydus image simultaneously with a single image reference.
-      merge_manifest: false
```
Comment (liubogithub) [inline misc/config/config.yaml.nydus.tmpl:53]: merge_manifest is also removed

File: pkg/driver/nydus/nydus.go [hunk 5]
```diff
@@ -30,12 +35,58 @@ import (
 	"github.com/goharbor/acceleration-service/pkg/utils"
 )
 
+type chunkDictInfo struct {
+	BootstrapPath string
+	Blobs         map[string]ocispec.Descriptor
```
Comment (liubogithub) [inline pkg/driver/nydus/nydus.go:43]: Why do we need `Blobs` given it's not really used?

File: pkg/driver/nydus/convert.go (hunk 1 above)
Comment (imeoer) [reply pkg/driver/nydus/convert.go]: Thanks for the review, this is a change of middle working commit.

File: pkg/driver/nydus/utils/archive.go (hunk 2 above)
Comment (imeoer) [reply pkg/driver/nydus/utils/archive.go]: Fixed

File: pkg/driver/nydus/utils/utils.go (hunk 3 above)
Comment (imeoer) [reply pkg/driver/nydus/utils/utils.go]: Fixed

File: pkg/driver/nydus/nydus.go (hunk 5 above)
Comment (imeoer) [reply pkg/driver/nydus/nydus.go]: Fixed, thanks.

**Review by liubogithub — APPROVED** [review]

**Review by liubogithub — APPROVED** [review]

**Pull request merged**